core/shared/src/main/scala/scalaz/zio/Retry.scala

```diff
@@ -0,0 +1,428 @@
+    def proj(state: State): S                = state
+    def update(e: E, s: State): IO[E, State] = update0(e, s)
+  }
+
+  /**
+   * A retry strategy that always fails.
+   */
+  final def never[E]: Retry[E, Unit] =
+    Retry[E, Unit](IO.unit, (e, _) => IO.fail(e))
+
+  /**
+   * A retry strategy that always succeeds.
+   */
+  final def always[E]: Retry[E, Unit] =
+    Retry[E, Unit](IO.unit, (e, _) => IO.unit)
+
+  /**
+   * A retry strategy that always succeeds with the specified constant state.
+   */
+  final def point[E, S](s: => S): Retry[E, S] =
+    Retry[E, S](IO.point(s), (_, s) => IO.now(s))
+
+  /**
+   * A retry strategy that always succeeds, collecting all errors into a list.
+   */
+  final def errors[E]: Retry[E, List[E]] =
+    Retry[E, List[E]](IO.now(Nil), (e, l) => IO.now(e :: l))
+
+  /**
+   * A retry strategy that always retries and counts the number of retries.
+   */
+  final def counted[E]: Retry[E, Int] =
+    Retry[E, Int](IO.now(0), (_, i) => IO.now(i + 1))
+
+  /**
+   * A retry strategy that always retries and computes the time since the
+   * beginning of the process.
+   */
+  final def elapsed[E]: Retry[E, Duration] = {
+    val nanoTime = IO.sync(System.nanoTime())
+
+    Retry[E, (Long, Long)](nanoTime.seq(IO.now(0L)), (_, t) => nanoTime.map(t2 => (t._1, t2 - t._1)))
+      .map(t => Duration(t._2, TimeUnit.NANOSECONDS))
+  }
+
+  /**
+   * A retry strategy that will keep retrying until the specified number of
+   * retries is reached.
+   */
+  final def retries[E](max: Int): Retry[E, Int] = counted.untilState(_ >= max)
+
+  /**
+   * A retry strategy that will keep retrying until the specified duration has
+   * elapsed.
+   */
+  final def duration[E](duration: Duration): Retry[E, Duration] = {
+    val nanos = duration.toNanos
+
+    elapsed.untilState(_.toNanos >= nanos)
+  }
+
+  /**
+   * A retry strategy that will always succeed, waiting the specified fixed
+   * duration between attempts.
+   */
+  final def fixed[E](duration: Duration): Retry[E, Int] =
+    counted.updated((_, _, io) => io.delay(duration))
+
+  final def backoff[E](base: Duration,
+                       sleep: (Int, Duration) => IO[E, Duration],
+                       cap: Option[Duration] = None): Retry[E, (Int, Duration)] = {
+    val up: ((Int, Duration)) => IO[E, (Int, Duration)] = {
+      case (n, d) => sleep(n, d).map(d0 => (n + 1, d0)).delay(cap.fold(d)(_.min(d)))
+    }
+    Retry[E, (Int, Duration)](IO.now((0, base)), (_, s) => up(s))
+  }
+
+  /**
+   * A retry strategy that will always succeed, but will wait a certain amount
+   * between retries, given by `base * factor.pow(n)`, where `n` is the
+   * number of retries so far.
+   */
+  final def exponentialBackoff[E](base: Duration,
+                                  factor: Double = 2.0,
+                                  cap: Option[Duration] = None): Retry[E, (Int, Duration)] =
+    backoff(base, (n, _) => IO.now(base * math.pow(factor, n.doubleValue)), cap)
```

Comment:
Same comment for here. `cap` does not need to exist because it can be layered in. I'd also just call this `exponential` so you can call the jitter one `exponentialJitter`.